portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java

```diff
@@ -0,0 +1,187 @@
+
+  // This list of all the time ranges will be sorted by start time ascending
+  // in add and delete methods. The time ranges stored in this list are
+  // pair-wise disjoint at any moment.
+  public List<TimeRange> allTimeRanges;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    // If the original allTimeRanges list is empty,
+    // this is the first time we add anything to the list,
+    // so simply add the new time range to the list and return.
+    if (allTimeRanges.isEmpty()) {
+      allTimeRanges.add(timeRange);
+      return;
+    }
+    ListIterator<TimeRange> iterator = allTimeRanges.listIterator();
+
+    // This variable represents the latest time range previously exmamined.
+    // Initially, this variable points to the time range we want to add.
+    // As the for loop iterates through all the current time ranges already existing
+    // in allTimeRanges, this lastExaminedTimeRange variable is the time range with which
+    // any existing time range merges with, if necessary.
+    TimeRange lastExaminedTimeRange = timeRange;
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+
+      // If the current range is completely contained by the lastExaminedTimeRange,
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        iterator.remove();
+        continue;
+      }
+
+      // The case for when two time ranges need to be merged.
+      // Similar to the case above, it will eventually be replaced so we can
+      // remove currentRange
+      if (lastExaminedTimeRange.overlaps(currentRange)) {
+        lastExaminedTimeRange = mergeTwoTimeRanges(currentRange, lastExaminedTimeRange);
+        iterator.remove();
+      } else if (currentRange.end().isAfter(lastExaminedTimeRange.end())) {
+        // This is the case that lastExaminedTimeRange and the current time range do not overlap.
+        // If the current range from the original list ends after the last examined time range,
+        // add the time range pointed to by lastExaminedTimeRange to the new list before the current
+        // range,and change the lastExaminedTimeRange to point to the current range.
+        iterator.previous();
+        iterator.add(lastExaminedTimeRange);
+        iterator.next();
+        lastExaminedTimeRange = currentRange;
+      }
```

Comment:
Should we consider the else case? It can get triggered for example if the internal list was just 1pm-2pm, and timeRange is 3pm-4pm.

Reply:
The else case means that the currentRange and lastExaminedTimeRange do not overlap and that lastExaminedTimeRange is strictly later than currentRange. On some further iteration the else if will trigger and insert it into the list once it comes before something in the list so we don't have to worry about the else case.